satellite/satellitedb/overlaycache.go

```diff
@@ -926,53 +926,66 @@ func (cache *overlaycache) AllPieceCounts(ctx context.Context) (_ map[storj.Node
 		}
 		pieceCounts[nodeID] = int(row.PieceCount)
 	}
+
 	return pieceCounts, nodeIDErrs.Err()
 }
 
 func (cache *overlaycache) UpdatePieceCounts(ctx context.Context, pieceCounts map[storj.NodeID]int) (err error) {
-	// TODO: add monkit stuff
 	defer mon.Task()(&ctx)(&err)
-
 	if len(pieceCounts) == 0 {
 		return nil
 	}
 
-	switch cache.db.Driver().(type) {
-	case *pq.Driver:
-		return Error.Wrap(cache.postgresUpdatePieceCounts(ctx, pieceCounts))
-	default:
-		return Error.Wrap(cache.sqliteUpdatePieceCounts(ctx, pieceCounts))
+	// TODO: pass in the apprioriate struct to database, rather than constructing it here
+	type NodeCount struct {
+		ID    storj.NodeID
+		Count int64
 	}
-}
-
-func (cache *overlaycache) postgresUpdatePieceCounts(ctx context.Context, pieceCounts map[storj.NodeID]int) (err error) {
-	defer mon.Task()(&ctx)(&err)
+	var counts []NodeCount
 
-	sqlQuery := `UPDATE nodes SET piece_count = newvals.piece_count FROM ( VALUES `
-	args := make([]interface{}, 0, len(pieceCounts)*2)
-	for nodeID, pieceCount := range pieceCounts {
-		sqlQuery += `(?::BYTEA, ?::BIGINT), `
-		args = append(args, nodeID, pieceCount)
-	}
-	sqlQuery = sqlQuery[:len(sqlQuery)-2]
-	// trim off the last comma+space
-	sqlQuery += `) newvals(nodeid, piece_count) WHERE nodes.id = newvals.nodeid`
-	_, err = cache.db.DB.ExecContext(ctx, cache.db.Rebind(sqlQuery), args...)
-	return err
-}
+	for nodeid, count := range pieceCounts {
+		counts = append(counts, NodeCount{
+			ID:    nodeid,
+			Count: int64(count),
+		})
+	}
+	sort.Slice(counts, func(i, k int) bool {
+		return counts[i].ID.Less(counts[k].ID)
+	})
 
-func (cache *overlaycache) sqliteUpdatePieceCounts(ctx context.Context, pieceCounts map[storj.NodeID]int) (err error) {
-	defer mon.Task()(&ctx)(&err)
+	switch t := cache.db.Driver().(type) {
+	case *sqlite3.SQLiteDriver:
+		err = cache.db.WithTx(ctx, func(ctx context.Context, tx *dbx.Tx) error {
+			query := tx.Rebind(`UPDATE nodes SET piece_count = ? WHERE id = ?`)
+			for _, count := range counts {
+				_, err := tx.Tx.ExecContext(ctx, query, count.Count, count.ID)
+				if err != nil {
+					return Error.Wrap(err)
```

Comment:
Wouldnt this mean that if a single update fails, we return out of the entire function and discard the other inserts.
And the piece count for some nodes could be updated by then, as i dont see that we rollback in an error case?

Reply:
I am not entirely up to speed how critical this number is..

Reply:
> Wouldnt this mean that if a single update fails, we return out of the entire function and discard the other inserts.

Yes, but this would also mean that there's something critically wrong already. Either we ended up creating piece count updates for storage nodes that aren't in our table. Or somebody deleted the node from the table... either would be problematic.

There of course could be an issue with database.

Of course, we don't care that much here, since it's for sqlite, which is mainly for testing.

> And the piece count for some nodes could be updated by then, as i dont see that we rollback in an error case?

When `WithTx` returns an error, it will rollback the transaction.

Reply:
Thanks, that should be ok then 👍